keps/sig-architecture/4330-compatibility-versions/README.md

```diff
@@ -0,0 +1,1301 @@
+Kubernetes binary version is working as desired before exposing any feature changes
+introduced in 1.31 to cluster users, and without writing and data to storage at
+newer API versions.
+
+This extra step increases the granularity of our upgrade sequence so that 
+(1) failures are more easily diagnosed (since we have more granular steps, we
+have slightly more data sheerly from which point of the upgrade lifecycle we
+achieved failure) and (2) failures are more easily auto-reverted by upgrade
+orchestration as we are taking smaller and more incremental steps forward,
+which means there is less to “undo” on a failure condition.
+
+that it would be possible to skip binary versions while still performing a
+stepwise upgrade of Kubernetes control-plane. For example:
+
+- (starting point) binary-version 1.28 (compat-version 1.28)
+- upgrade binary-version to 1.31 (emulation-version stays at 1.28 - this is our skip-level binary upgrade)
+- keep binary-version 1.31 while upgrading emulation-version to 1.29 (stepwise upgrade of emulation version)
+- keep binary-version 1.31 while upgrading emulation-version to 1.30 (stepwise upgrade of emulation version)
+- keep binary-version 1.31 while upgrading emulation-version to 1.31 (stepwise upgrade of emulation version)
+
+Benefits to upgrading binary version independent of emulation version:
+
+- A skip-version binary upgrade that transitions through emulation versions has
+  higher probability of bugs in those intermediate binary versions already being
+  fixed.
+- During an upgrade, it is possible upgrade the new binary version while the
+  emulation version is fixed (e.g. `binaryVersion: 1.28` -> `{binaryVersion: 1.29, emulationVersion: 1.28}`).
+  This allow differences in the binary (bugs fixed or introduced, performance
+  changes, ...) to be introduced into a cluster and verified safe before
+  allowing access to the new APIs and features new version. Once the binary
+  version has been deamed safe, the emulation version can then be upgraded.
+- Any upgrade system that successfully detects failures between upgrade steps
+  can report which upgrade step failed. This makes it easier to diagnose the
+  failures, because there are fewer possible causes of the failure. (There's a
+  huge difference between "A cluster upgrade failed" and "A cluster upgrade
+  failed when only the apiserver's binary version changed").
+- For upgrades where multiple failures occur, this increases the odds those
+  failures are split across steps. An upgrade system that is able to pause after
+  a step where failures are detected allow for failures at that step to be
+  addressed before proceeding to subsequent steps. These failures can be
+  addressed without the disruption and "noise" from failures in subsequent
+  steps.
+- A emulation version rollback can be performed without changing binary version.
+
+A dedicated `--min-compatibility-version` flag provides direct control of when
```

Comment:
Ok, this is good hear but something like this probably needs to be in the summary too. At least a one-liner giving the primary use case for min-compatibility-version vs emulation-version. Otherwise in the summary you just sort of have to "wait and see".

Reply:
SG. I've tried to boil it down to the key properties in the summary.